debug/debugger.go

```diff
@@ -0,0 +1,859 @@
+
+	// AddBreakpoint sets a breakpoint at the given location.
+	AddBreakpoint(loc location.Location) (Breakpoint, error)
+
+	// RemoveBreakpoint removes a given breakpoint.
+	// The removed breakpoint is returned. If the breakpoint does not exist, nil is returned.
+	RemoveBreakpoint(ID BreakpointID) (Breakpoint, error)
+
+	// ClearBreakpoints removes all breakpoints.
+	ClearBreakpoints() error
+
+	// StackTrace returns the StackTrace for the thread with the given ID.
+	// The stack trace is ordered from the most recent frame to the least recent frame.
+	StackTrace(threadID ThreadID) (StackTrace, error)
+
+	// Scopes returns the Scope list for the frame with the given ID.
+	Scopes(frameID FrameID) ([]Scope, error)
+
+	// Variables returns the Variable list for the given reference.
+	Variables(varRef VarRef) ([]Variable, error)
+
+	// Terminate stops all threads in the session.
+	Terminate() error
+
+	// TODO: Add Stop(ThreadID) func for stopping (pausing) a thread's execution.
+}
+
+type printHook struct {
+	prnt.Hook
+	d *debugger
+}
+
+func (h *printHook) Print(_ prnt.Context, str string) error {
+	if h == nil || h.d == nil {
+		return nil
+	}
+	h.d.eventHandler(Event{Type: StdoutEventType, Message: str})
+	return nil
+}
+
+type DebuggerOption func(*debugger)
+
+func NewDebugger(options ...DebuggerOption) Debugger {
+	return newDebugger(options...)
+}
+
+func newDebugger(options ...DebuggerOption) *debugger {
+	d := &debugger{
+		eventHandler: newNopEventHandler(),
+		logger:       logging.NewNoOpLogger(),
+	}
+	d.printHook = &printHook{d: d}
+
+	for _, option := range options {
+		option(d)
+	}
+
+	return d
+}
+
+func SetLogger(logger logging.Logger) DebuggerOption {
+	return func(d *debugger) {
+		d.logger = logger
+	}
+}
+
+func SetEventHandler(handler EventHandler) DebuggerOption {
+	return func(d *debugger) {
+		d.eventHandler = handler
+	}
+}
+
+type LaunchEvalProperties struct {
+	LaunchProperties
+	Query     string
+	Input     interface{}
+	InputPath string
+}
+
+type LaunchTestProperties struct {
+	LaunchProperties
+	Run string `json:"run"`
+}
+
+type LaunchProperties struct {
+	BundlePaths  []string
+	DataPaths    []string
+	StopOnResult bool
+	StopOnEntry  bool
+	StopOnFail   bool
+	EnablePrint  bool
+	SkipOps      []topdown.Op
+	RuleIndexing bool
+}
+
+func (lp LaunchProperties) String() string {
+	b, err := json.Marshal(lp)
+	if err != nil {
+		return "{}"
+	}
+	return string(b)
+}
+
+func (d *debugger) LaunchEval(ctx context.Context, props LaunchEvalProperties) (Session, error) {
+	store := inmem.New()
+	txn, err := store.NewTransaction(ctx, storage.TransactionParams{Write: true})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create store transaction: %v", err)
+	}
+
+	regoArgs := []func(*rego.Rego){
+		rego.Query(props.Query),
+		rego.Store(store),
+		rego.Transaction(txn),
+	}
+
+	if props.SkipOps == nil {
+		props.SkipOps = []topdown.Op{topdown.IndexOp, topdown.RedoOp, topdown.SaveOp, topdown.UnifyOp}
```

Comment:
Because they are generally not interesting to halt execution on, and will always be preceded by an `EvalOp`, `FailOp`, or `ExitOp`, which we do halt on. A user has the option to declare a different skip-list if this isn't suitable for the edge case they're investigating.